File: chess_engine/src/book.cpp
// book.cpp — Polyglot opening book (stub).
//
// A real Polyglot reader requires the 781-entry Polyglot Random64 constant
// table to compute book-compatible Zobrist keys. That table is not embedded
// in this release. Calling find_move() always returns MOVE_NONE; the engine
// falls back to its search. The Python implementation in chess_engine/book.py
// has a working Polyglot reader if you need book moves now.

#include "book.h"
#include "board.h"

#include <fstream>
#include <iostream>

bool OpeningBook::load(const std::string& path) {
    std::ifstream f(path, std::ios::binary);
    if (!f) {
        std::cerr << "info string book: failed to open " << path << '\n';
        loaded_ = false;
        return false;
    }
    // Sanity: a polyglot file is a multiple of 16 bytes.
    f.seekg(0, std::ios::end);
    auto sz = f.tellg();
    if (sz < 16 || (sz % 16) != 0) {
        std::cerr << "info string book: not a valid polyglot file\n";
        loaded_ = false;
        return false;
    }
    std::cerr << "info string book: loaded " << path
              << " (" << (sz / 16) << " entries) but key generation is not "
              << "implemented; book is currently inert\n";
    loaded_ = true;
    return true;
}

Move OpeningBook::find_move(const Board& /*b*/) {
    // TODO: implement polyglot key (requires the 781 Random64 constants).
    return MOVE_NONE;
}

void OpeningBook::close() { loaded_ = false; }

File: chess_engine/book.py
"""Polyglot opening book reader (thin wrapper over python-chess).

Polyglot books store positions by Zobrist hash and a list of moves with
weights. We pick a move proportionally to its weight so games don't all
collapse onto the same line.
"""

import os
import random

import chess
import chess.polyglot


class OpeningBook:
    def __init__(self, path=None):
        self.path = path
        self.reader = None
        if path and os.path.exists(path):
            try:
                self.reader = chess.polyglot.open_reader(path)
            except Exception:
                self.reader = None

    def find_move(self, board, weighted=True):
        if self.reader is None:
            return None
        try:
            entries = list(self.reader.find_all(board))
        except Exception:
            return None
        if not entries:
            return None
        if not weighted:
            return entries[0].move
        total = sum(e.weight for e in entries)
        if total <= 0:
            return entries[0].move
        pick = random.randint(1, total)
        running = 0
        for e in entries:
            running += e.weight
            if running >= pick:
                return e.move
        return entries[0].move

    def close(self):
        if self.reader is not None:
            self.reader.close()
            self.reader = None
